File: 10.hooks/qemu.py
#!/usr/bin/python3
import os
import sys
import time
from datetime import datetime
import psutil

guest_name = sys.argv[1]
libvirt_task = sys.argv[2]

enabled_video_processor = False

passthrough_vm_names = ["win11", "win10", "ubuntu", "manjaro", "win7"]

need_passthrough = False
if guest_name in passthrough_vm_names:
    need_passthrough = True


def check_process_exists(process_name):
    for process in psutil.process_iter(["name"]):
        if process.info["name"] == process_name:
            return process.pid
    return -1


def open_gui():
    return
    time.sleep(5)
    print("OpenGUI")
    os.system("startx")


def close_gui():
    return
    print("CloseGUI")
    exist = False

    # gui_process_name = "xinit"
    gui_process_name = "Xorg"

    # 检查进程是否存在
    pid = check_process_exists(gui_process_name)
    if pid > 0:
        print(f"Process {gui_process_name} exist")
        os.kill(pid, 15)
        # os.system("pkill Xorg")
        wait_count = 0
        while True:
            wait_count += 1
            # 检查Process是否存在
            pid = check_process_exists(gui_process_name)
            if pid > 0:
                time.sleep(1)
                print(f"Waiting for {gui_process_name}")
                continue
            else:
                print(f"Process {gui_process_name} terminated")
                time.sleep(3)
                break
    else:
        print(f"Process {gui_process_name} not exist")
        time.sleep(0.5)


def create_snapshot(guest_name):
    dataset = f"/jdata/vdisk/{guest_name}"
    if os.path.exists(dataset):
        print(f"路径 {dataset} 存在")
        cmd = f"/jdata/develop/AutoSnapshot/SnapshotHelper.py snap {dataset}"
        print(cmd)
        os.system(cmd)
    else:
        print(f"路径 {dataset} 不存在")


def on_prepare():
    # check nvidia
    if check_nvidia_module_loaded():
        print("已加载含nvidia的内核模块")
    else:
        print("未加载含nvidia的内核模块")

    # 暂时去掉开机快照，防止创建不必要的快照
    # create_snapshot(guest_name)

    if need_passthrough:
        close_gui()
        # os.system("modprobe -r nvidia_drm")
        # os.system("modporbe -r nvidia_uvm")
        # os.system("modprobe -r nvidia")


def on_release():
    if need_passthrough:
        # os.system("modprobe nvidia")
        # os.system("modporbe nvidia_uvm")
        # os.system("modprobe nvidia_drm")
        open_gui()


def qemu_process():
    print(
        f"{datetime.now()}: {guest_name} {libvirt_task}, passthrough:{need_passthrough}"
    )

    if libvirt_task.__contains__("prepare"):
        on_prepare()
    elif libvirt_task.__contains__("release"):
        on_release()


def check_nvidia_module_loaded():
    try:
        with open("/proc/modules", "r") as f:
            content = f.read()
            lines = content.splitlines()
            for line in lines:
                module_name = line.split()[0]
                if "nvidia" in module_name.lower():
                    return True
            return False
    except FileNotFoundError:
        print("/proc/modules文件不存在，可能不是Linux系统或者系统存在异常")
        return False


qemu_process()
print("-----------------------end---------------------------")

File: 10.hooks/src/qemu.ts
#!/usr/bin/env bun

import { $ } from "bun";
import { existsSync, readFileSync, appendFileSync } from "fs";
import { dirname } from "path";

// 切换到脚本所在目录
const scriptDir = dirname(import.meta.path.replace("file://", ""));
process.chdir(scriptDir);

// 从命令行参数获取虚拟机名称和 libvirt 任务
const guestName = process.argv[2];
const libvirtTask = process.argv[3];

if (!guestName || !libvirtTask) {
  console.error("Usage: qemu.ts <guest_name> <libvirt_task>");
  process.exit(1);
}

// 类型断言：经过上面的检查，这两个变量肯定不是 undefined
const vmName: string = guestName;
const taskName: string = libvirtTask;

// 日志文件路径
const LOG_FILE = "/var/log/libvirt/hooks.log";

/**
 * 写入日志到文件
 * @param message 日志消息
 */
function logToFile(message: string): void {
  try {
    appendFileSync(LOG_FILE, message + "\n");
  } catch (error) {
    // 如果无法写入日志文件，忽略错误（可能是权限问题）
  }
}

/**
 * 同时输出到控制台和日志文件
 * @param message 消息
 */
function log(message: string): void {
  console.log(message);
  logToFile(message);
}

const enabledVideoProcessor = false;

// 需要 GPU 直通的虚拟机列表
const passthroughVmNames = ["win11", "win10", "ubuntu", "manjaro", "win7"];

// 判断是否需要 GPU 直通
const needPassthrough = passthroughVmNames.includes(vmName);

/**
 * 检查指定进程是否存在
 * @param processName 进程名称
 * @returns 进程 PID，不存在返回 -1
 */
async function checkProcessExists(processName: string): Promise<number> {
  try {
    // 使用 pgrep 查找进程
    const result = await $`pgrep -x ${processName}`.quiet();
    const output = result.stdout.toString().trim();

    if (output) {
      const firstLine = output.split('\n')[0];
      if (firstLine) {
        const pid = parseInt(firstLine);
        return pid;
      }
    }
    return -1;
  } catch (error) {
    // pgrep 找不到进程时会返回非零退出码
    return -1;
  }
}

/**
 * 打开 GUI
 */
async function openGui(): Promise<void> {
  return; // 当前已禁用

  await sleep(5000);
  log("OpenGUI");
  await $`startx`.quiet();
}

/**
 * 关闭 GUI
 */
async function closeGui(): Promise<void> {
  return; // 当前已禁用

  log("CloseGUI");

  // const guiProcessName = "xinit";
  const guiProcessName = "Xorg";

  // 检查进程是否存在
  let pid = await checkProcessExists(guiProcessName);

  if (pid > 0) {
    log(`Process ${guiProcessName} exist`);

    // 发送 SIGTERM 信号 (15)
    try {
      process.kill(pid, 15);
    } catch (error) {
      logToFile(`Failed to kill process ${pid}: ${error}`);
    }

    let waitCount = 0;
    while (true) {
      waitCount++;

      // 检查进程是否存在
      pid = await checkProcessExists(guiProcessName);

      if (pid > 0) {
        await sleep(1000);
        log(`Waiting for ${guiProcessName}`);
        continue;
      } else {
        log(`Process ${guiProcessName} terminated`);
        await sleep(3000);
        break;
      }
    }
  } else {
    log(`Process ${guiProcessName} not exist`);
    await sleep(500);
  }
}

/**
 * 创建快照
 * @param guestName 虚拟机名称
 */
async function createSnapshot(guestName: string): Promise<void> {
  const dataset = `/jdata/vdisk/${guestName}`;

  if (existsSync(dataset)) {
    log(`路径 ${dataset} 存在`);
    const cmd = `/jdata/develop/AutoSnapshot/SnapshotHelper.py snap ${dataset}`;
    log(cmd);
    await $`${cmd}`.quiet();
  } else {
    log(`路径 ${dataset} 不存在`);
  }
}

/**
 * 检查 nvidia 内核模块是否已加载
 * @returns 是否已加载
 */
function checkNvidiaModuleLoaded(): boolean {
  try {
    const content = readFileSync("/proc/modules", "utf-8");
    const lines = content.split('\n');

    for (const line of lines) {
      if (line.trim() === '') continue;

      const moduleName = line.split(/\s+/)[0];
      if (moduleName && moduleName.toLowerCase().includes("nvidia")) {
        return true;
      }
    }
    return false;
  } catch (error) {
    log("/proc/modules文件不存在，可能不是Linux系统或者系统存在异常");
    return false;
  }
}

/**
 * 准备阶段处理 (虚拟机启动前)
 */
async function onPrepare(): Promise<void> {
  // 检查 nvidia 模块
  if (checkNvidiaModuleLoaded()) {
    log("已加载含nvidia的内核模块");
  } else {
    log("未加载含nvidia的内核模块");
  }

  // 暂时去掉开机快照，防止创建不必要的快照
  // await createSnapshot(guestName);

  if (needPassthrough) {
    await closeGui();
    // await $`modprobe -r nvidia_drm`.quiet();
    // await $`modprobe -r nvidia_uvm`.quiet();
    // await $`modprobe -r nvidia`.quiet();
  }
}

/**
 * 释放阶段处理 (虚拟机关闭后)
 */
async function onRelease(): Promise<void> {
  if (needPassthrough) {
    // await $`modprobe nvidia`.quiet();
    // await $`modprobe nvidia_uvm`.quiet();
    // await $`modprobe nvidia_drm`.quiet();
    await openGui();
  }
}

/**
 * QEMU 进程处理主函数
 */
async function qemuProcess(): Promise<void> {
  const now = new Date().toISOString();
  const message = `${now}: ${vmName} ${taskName}, passthrough:${needPassthrough}`;

  // 写入日志文件头部
  logToFile("");
  logToFile(`Date: ${new Date().toLocaleString('zh-CN', { timeZone: 'Asia/Shanghai' })}`);
  logToFile(`Command: ./qemu.ts "${vmName}" "${taskName}"`);

  log(message);

  if (taskName.includes("prepare")) {
    await onPrepare();
  } else if (taskName.includes("release")) {
    await onRelease();
  }
}

/**
 * 睡眠函数
 * @param ms 毫秒数
 */
function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// 主执行流程
try {
  await qemuProcess();
  log("-----------------------end---------------------------");
  logToFile(""); // 添加空行分隔
} catch (error) {
  const errorMsg = `Error occurred: ${error}`;
  console.error(errorMsg);
  logToFile(errorMsg);
  logToFile("");
  process.exit(1);
}
